src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1075,8 +1080,47 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
             }
 
             diagnostics.AddRangeAndFree(localDiagnostics);
-            var boundDeclType = new BoundTypeExpression(typeSyntax, aliasOpt, inferredType: isVar, type: declTypeOpt.TypeSymbol);
+
+            BoundTypeExpression boundDeclType = null;
+
+            if (bindType)
+            {
+                var invalidDimensions = ArrayBuilder<BoundExpression>.GetInstance();
+
+                // We don't want to report any errors from this, as the syntax is invalid anyway
+                var dummyDiagnostics = DiagnosticBag.GetInstance();
+
+                GetInvalidDimensions(typeSyntax, invalidDimensions, dummyDiagnostics);
+
+                dummyDiagnostics.Free();
+
+                boundDeclType = new BoundTypeExpression(typeSyntax, aliasOpt, inferredType: isVar, dimensionsOpt: invalidDimensions.ToImmutableAndFree(), type: declTypeOpt.TypeSymbol);
+            }
+
             return new BoundLocalDeclaration(associatedSyntaxNode, localSymbol, boundDeclType, initializerOpt, arguments, hasErrors);
+
+            void GetInvalidDimensions(TypeSyntax node, ArrayBuilder<BoundExpression> invalidDimensions, DiagnosticBag diagnosticsBag)
```

Comment:
>GetInvalidDimensions [](start = 17, length = 20)

According to our coding conventions, names of local functions shouldn't start with a capital letter.